### ISSUE TITLE
Inconsistent output when passing c for IBP vs when just working with lb and ub (for a classifier)

### ISSUE THREAD

**Issue by aknirala** [Verified-Intelligence/auto_LiRPA#88]
Context: I am working with an image classifier which classifies an input to `num_classes`. Also I am only using `IBP` and nothing from `CROWN`.

As per my understanding, while calculating the output for a bounded input, one may or may not pass matrix c as (here c is passed but it can also be None):
```
lb, ub = model(method_opt="compute_bounds", x=x, IBP=True, 
                C=get_spec_matrix(images, targets, num_classes), 
                method='IBP', no_replicas=True)
rob_correct = len(lb) - torch.sum((lb < 0).any(dim=1)).item()
```
As per my understanding when c is passed, the way it is constructed, it subtracts the upper bound for all the logits from lower bound of the correct logit. if all the values of lb are positive this indicates correct classification.

Now, if c is not passed, we simply get lower and upper bound for the logits. Thus, if I construct rob_logits as below (i.e, it is same as ub but I am replacing correct logit with lower bound):
```
lb_simp, ub_simp = model(method_opt="compute_bounds", x=x, IBP=True, 
                C=None, 
                method='IBP', no_replicas=True)
rob_logits = ub_simp.clone()
batch_indices = torch.arange(rob_logits.size(0))  # [0, 1, ..., B-1]
rob_logits[batch_indices, targets] = lb_simp[batch_indices, targets]
rob_correct_simp = (rob_logits.data.max(1)[1] == targets).float().sum()
```
and than simply check if highest logit index matches the correct targets/label (as done for natural classification),  than also it shall give the same value (i.e, should classify the input correctly or incorrectly). However, I get dramatically different results.

**Q1:** I would really appreciate if someone could explain this to me please.

Here is a code to reproduce the issue:
```
from auto_LiRPA import BoundedModule
from auto_LiRPA.perturbations import PerturbationLpNorm
from auto_LiRPA import BoundedTensor
from auto_LiRPA.utils import get_spec_matrix

from datasets import CIFAR10
from models import *
import torchvision.transforms as transforms

model_ori = wide_resnet_8()
chkpt = torch.load("cifar_wide_resnet_8_160epochs/ckpt_160")
model_ori.load_state_dict(chkpt['state_dict'])
print("Model loaded!!")
device =  torch.device("cuda")

#Load the CIFAR data
cifar10_mean = [0.4914, 0.4822, 0.4465]
cifar10_std = [0.2023, 0.1994, 0.2010]
use_index = False
batch_size = test_batch_size = 32
num_classes = 10
mean = torch.tensor(cifar10_mean)
std = torch.tensor(cifar10_std)
dummy_input = torch.randn(2, 3, 32, 32)
normalize = transforms.Normalize(mean = mean, std = std)
transform = transforms.Compose([
            transforms.RandomHorizontalFlip(),
            transforms.RandomCrop(32, 2, padding_mode='edge'),
            transforms.ToTensor(),
            normalize])
transform_test = transforms.Compose([transforms.ToTensor(), normalize])

train_data = CIFAR10('./data', train=True, download=True, transform=transform, use_index=use_index)
test_data = CIFAR10('./data', train=False, download=True, transform=transform_test, use_index=use_index)
train_loader = torch.utils.data.DataLoader(train_data, batch_size=batch_size, shuffle=True, pin_memory=True, num_workers=0)
test_loader = torch.utils.data.DataLoader(test_data, batch_size=test_batch_size, pin_memory=True, num_workers=0)

train_loader.mean = test_loader.mean = mean
train_loader.std = test_loader.std = std  

for loader in [train_loader, test_loader]:
    loader.mean, loader.std = mean, std
    loader.data_max = torch.reshape((1. - mean) / std, (1, -1, 1, 1)).to(device)
    loader.data_min = torch.reshape((0. - mean) / std, (1, -1, 1, 1)).to(device)
    loader.mean     = torch.reshape(mean, (1, -1, 1, 1)).to(device)
    loader.std      = torch.reshape(std,  (1, -1, 1, 1)).to(device)
dummy_input = dummy_input.to(device)

model_ori.eval()
model = BoundedModule(model_ori, dummy_input, bound_opts={'relu':None}, device=device)
model.eval()
print("Bounded model created")
print(f"test_loader: {len(test_loader)}; test_data: {len(test_data)}")

for i, (images, targets) in enumerate(test_loader): break
images = images.to(device, non_blocking=True)
targets = targets.to(device, non_blocking=True)
op = model(images)
eps = (8/255)/test_loader.std
data_ub = torch.min(images + eps, test_loader.data_max)
data_lb = torch.max(images - eps, test_loader.data_min)
ptb = PerturbationLpNorm(norm=np.inf, eps=eps, x_L=data_lb, x_U=data_ub)
x = BoundedTensor(images, ptb)
lb, ub = model(method_opt="compute_bounds", x=x, IBP=True, 
                C=get_spec_matrix(images, targets, num_classes), 
                method='IBP', no_replicas=True)
rob_correct = len(lb) - torch.sum((lb < 0).any(dim=1)).item()
nat_correct =  float((op.data.max(1)[1] == targets).float().sum())
print(f"rob_correct: {rob_correct}, nat_correct: {nat_correct}")

#Now do it with lb and ub without using c
lb_simp, ub_simp = model(method_opt="compute_bounds", x=x, IBP=True, 
                C=None, 
                method='IBP', no_replicas=True)
rob_logits = ub_simp.clone()
batch_indices = torch.arange(rob_logits.size(0))  # [0, 1, ..., B-1]
rob_logits[batch_indices, targets] = lb_simp[batch_indices, targets]
rob_correct_simp = (rob_logits.data.max(1)[1] == targets).float().sum()
print("rob_correct_simp is:", rob_correct_simp)

torch.set_printoptions(linewidth=200)
print("lb: ", lb[:10])
print("ub: ", ub[:10])
print("targets: ", targets[:10])
print("lb_simp: ", lb_simp[:10])
print("ub_simp: ", ub_simp[:10])
print("rob_logits: ", rob_logits[:10])
```

The output of the code is: **Q2:** Shouldn't I should be able to calculate values in lb and ub based on values in lb_simp, ub_simp
```
Bounded model created
test_loader: 313; test_data: 10000
rob_correct: 10, nat_correct: 16.0
rob_correct_simp is: tensor(3., device='cuda:0')
lb:  tensor([[ 1.6651,  1.5072,  0.2658, -0.0886,  0.0205, -0.1962,  0.3022,  0.8281,  2.6911],
        [ 0.3167,  0.7252,  3.3504,  5.1140,  3.5397,  6.9434,  7.6746,  4.1210,  0.0493],
        [ 0.7189,  0.7176,  2.7040,  3.8653,  2.9957,  4.4178,  5.8638,  1.9632,  0.1322],
        [-0.0441,  0.6871,  0.8121,  0.6762,  2.3634,  3.3583,  0.9555, -0.8618, -0.1825],
        [ 2.7679,  4.0541,  0.4940,  0.5593,  0.0858,  0.5801,  1.4738,  2.4863,  3.3999],
        [ 2.9187,  2.7902,  0.7970,  0.9646,  0.3384,  0.9856,  1.1474,  4.0121,  3.1826],
        [ 1.5651,  1.5391, -0.0299,  1.9506, -0.0146,  1.4277,  0.7965,  6.3110,  1.8962],
        [ 2.4560,  3.4057,  0.5354,  0.3263,  0.1869,  0.3471,  0.6425,  2.4301,  2.6861],
        [ 1.4035,  3.7926,  0.0984,  0.2466, -0.3139,  0.5303,  0.2414,  1.7839,  4.3225],
        [-0.6876,  0.9395,  1.8230,  1.1874,  2.4904,  2.3266,  1.5384, -0.5612, -0.4333]], device='cuda:0', grad_fn=<SliceBackward0>)
ub:  tensor([[ 1.9406e+00,  4.3600e+00,  4.5425e-01,  2.5234e-01,  2.0818e-02,  9.9224e-02,  1.1805e+00,  1.2297e+00,  5.0661e+00],
        [ 4.3340e-01,  1.1151e+00,  5.5485e+00,  8.9779e+00,  5.1877e+00,  1.3127e+01,  1.3319e+01,  6.2041e+00,  5.5888e-01],
        [ 7.4880e-01,  9.4764e-01,  5.7232e+00,  9.5087e+00,  5.7708e+00,  1.3861e+01,  1.4027e+01,  5.7488e+00,  9.1278e-01],
        [ 1.4957e+00,  2.6881e+00,  3.4659e+00,  3.1276e+00,  9.6049e+00,  8.8092e+00,  3.6502e+00, -6.2698e-01,  2.2305e+00],
        [ 3.4199e+00,  5.0557e+00,  8.4681e-01,  8.8338e-01,  4.0407e-01,  9.0436e-01,  1.9829e+00,  3.8109e+00,  4.4607e+00],
        [ 3.0463e+00,  4.0609e+00,  8.6841e-01,  1.1059e+00,  4.5171e-01,  1.1362e+00,  1.2784e+00,  5.2994e+00,  4.3067e+00],
        [ 2.8468e+00,  5.2550e+00,  1.6197e+00,  4.5820e+00,  1.6781e+00,  5.2089e+00,  3.1260e+00,  2.1155e+01,  2.5100e+00],
        [ 2.7948e+00,  5.0415e+00,  7.1456e-01,  5.2370e-01,  4.2275e-01,  5.4789e-01,  9.3421e-01,  3.4651e+00,  3.7279e+00],
        [ 2.0181e+00,  6.3266e+00,  4.8996e-01,  5.5837e-01,  2.0579e-02,  8.5607e-01,  5.0726e-01,  2.9163e+00,  6.1326e+00],
        [ 4.7436e-01,  2.9700e+00,  4.5813e+00,  3.0115e+00,  7.9496e+00,  6.8502e+00,  5.0361e+00,  4.3666e-01, -8.5516e-02]], device='cuda:0', grad_fn=<SliceBackward0>)
targets:  tensor([3, 8, 8, 0, 6, 6, 1, 6, 3, 1], device='cuda:0')
lb_simp:  tensor([[ -1.6582,  -3.9816,  -0.1580,   0.2296,   0.0837,   0.2088,   0.2242,  -0.8777,  -0.9291,  -4.6783],
        [  1.3768,   0.7872,  -3.2585,  -6.5261,  -2.9743, -10.5609, -10.8374,  -3.9559,   1.7220,   1.3767],
        [  0.6328,   0.5194,  -3.4288,  -6.9341,  -3.5609, -11.1001, -11.3511,  -3.3949,   1.3597,   0.7911],
        [  0.0869,  -0.5962,  -1.6281,  -2.2928,  -1.9998,  -8.2055,  -7.5194,  -2.4868,   0.7847,  -1.2787],
        [ -2.2759,  -3.8882,   0.2534,   0.1919,   0.6851,   0.1709,   0.8710,  -0.8462,  -2.6359,  -3.2942],
        [ -1.9972,  -2.8713,   0.1714,  -0.0552,   0.5968,  -0.0843,   0.9883,  -0.2182,  -4.1755,  -3.1258],
        [ -1.5170,   0.7562,  -3.5928,   0.0203,  -2.9644,  -0.0364,  -3.5210,  -1.5602, -19.2146,  -1.6751],
        [ -1.9817,  -4.1145,   0.0749,   0.2806,   0.3597,   0.2567,   0.6545,  -0.1204,  -2.5868,  -2.8240],
        [ -1.2806,  -5.5010,   0.1329,   0.4562,   0.0586,   0.4988,  -0.3724,   0.0982,  -2.1283,  -5.3245],
        [  0.1882,   0.1446,  -2.0635,  -3.6561,  -2.1431,  -6.8873,  -5.8382,  -4.1582,   0.1246,   0.2744]], device='cuda:0', grad_fn=<SliceBackward0>)
ub_simp:  tensor([[-1.0395, -0.9777,  0.3459,  0.6784,  0.6605,  0.6579,  0.7808,  0.3030, -0.2207, -2.1709],
        [ 2.5566,  2.0559, -0.9569, -2.8824, -1.0696, -4.8261, -5.4728, -1.6858,  2.9615,  2.6986],
        [ 2.4872,  2.4031, -0.4106, -1.8521, -0.6179, -2.5910, -3.9517,  0.2707,  3.2281,  2.7517],
        [ 1.7558,  0.9873,  0.0956, -0.1425,  0.0387, -1.9201, -2.8054, -0.2763,  2.5468,  1.0734],
        [-1.7779, -3.0877,  0.5397,  0.4994,  0.9590,  0.4786,  1.2629, -0.4766, -1.5274, -2.4326],
        [-1.7266, -1.7387,  0.4044,  0.2260,  0.8543,  0.2037,  1.2529,  0.0336, -2.8947, -2.1222],
        [ 0.5933,  2.7321,  0.2870,  1.8782, -0.0799,  1.8612,  0.3727,  1.1261, -4.7630,  0.7572],
        [-1.6010, -2.6645,  0.3433,  0.5375,  0.6988,  0.5165,  1.0136,  0.2119, -1.6403, -1.9218],
        [-0.6762, -3.1535,  0.7435,  1.0086,  0.6012,  1.2593,  0.4509,  0.6179, -1.1071, -3.6657],
        [ 1.7735,  1.6038, -0.0976, -0.9998, -0.3073, -1.8043, -1.5902, -0.6678,  1.7483,  1.9928]], device='cuda:0', grad_fn=<SliceBackward0>)
rob_logits:  tensor([[-1.0395, -0.9777,  0.3459,  0.2296,  0.6605,  0.6579,  0.7808,  0.3030, -0.2207, -2.1709],
        [ 2.5566,  2.0559, -0.9569, -2.8824, -1.0696, -4.8261, -5.4728, -1.6858,  1.7220,  2.6986],
        [ 2.4872,  2.4031, -0.4106, -1.8521, -0.6179, -2.5910, -3.9517,  0.2707,  1.3597,  2.7517],
        [ 0.0869,  0.9873,  0.0956, -0.1425,  0.0387, -1.9201, -2.8054, -0.2763,  2.5468,  1.0734],
        [-1.7779, -3.0877,  0.5397,  0.4994,  0.9590,  0.4786,  0.8710, -0.4766, -1.5274, -2.4326],
        [-1.7266, -1.7387,  0.4044,  0.2260,  0.8543,  0.2037,  0.9883,  0.0336, -2.8947, -2.1222],
        [ 0.5933,  0.7562,  0.2870,  1.8782, -0.0799,  1.8612,  0.3727,  1.1261, -4.7630,  0.7572],
        [-1.6010, -2.6645,  0.3433,  0.5375,  0.6988,  0.5165,  0.6545,  0.2119, -1.6403, -1.9218],
        [-0.6762, -3.1535,  0.7435,  0.4562,  0.6012,  1.2593,  0.4509,  0.6179, -1.1071, -3.6657],
        [ 1.7735,  0.1446, -0.0976, -0.9998, -0.3073, -1.8043, -1.5902, -0.6678,  1.7483,  1.9928]], device='cuda:0', grad_fn=<SliceBackward0>)
```

And here is a [colab](https://colab.research.google.com/drive/1X2GIDs5XvPO3HEj2zF8s8SPaJvTsEMuM#scrollTo=-_dX4guzZxbM) notebook with issue reproduced.

**Update** If after: 
`lb_simp, ub_simp = model.compute_bounds(x=x, IBP=True, C=None, method='IBP') `
I use: 
`lb, ub = model.compute_bounds(x=None, IBP=True,  C=c, method='IBP')`
Than lb and ub are calculated as expected. Why?

**Comment by shizhouxing** [comment]
Hi @aknirala, 

Passing `c` to IBP is recommended, as it typically yields tighter bounds. Post-processing (i.e., multiplying some `c` matrix) IBP bounds computed without `c` is not equivalent to the former one, as IBP losts the information on `c` and typically produces looser bounds. 

>Update If after:
>lb_simp, ub_simp = model.compute_bounds(x=x, IBP=True, C=None, method='IBP') 
>I use:
>lb, ub = model.compute_bounds(x=None, IBP=True,  C=c, method='IBP')
>Than lb and ub are calculated as expected. Why?

In this case, the second `compute_bounds` may reuse cached results from the first `compute_bounds`, as a new `x` is not provided. It could be equivalent to post-processing IBP bounds computed without `c`.
